### PR TITLE
Add documentation for speaker embeddings task and fix audio length bug

### LINKED ISSUE #91: Bug: extract_speaker_embeddings_from_audios only works on list of audio clips of the same size
### Description

`extract_speaker_embeddings_from_audios` only works on list of audio clips of the same size

### Steps to Reproduce

```
from senselab.audio.data_structures.audio import Audio
from senselab.audio.tasks.speaker_embeddings.api import extract_speaker_embeddings_from_audios
from senselab.utils.data_structures.model import SpeechBrainModel
from senselab.audio.tasks.preprocessing.preprocessing import downmix_audios_to_mono
from senselab.audio.tasks.preprocessing.preprocessing import resample_audios

MONO_AUDIO_PATH = "src/tests/data_for_testing/audio_48khz_mono_16bits.wav"
STEREO_AUDIO_PATH = "src/tests/data_for_testing/audio_48khz_stereo_16bits.wav"

audio1 = Audio.from_filepath(MONO_AUDIO_PATH)
audio2 = Audio.from_filepath(STEREO_AUDIO_PATH)
audio2 = downmix_audios_to_mono([audio2])[0]
[audio1, audio2] = resample_audios([audio1, audio2], resample_rate=16000)

ecapa_model = SpeechBrainModel(path_or_uri="speechbrain/spkrec-ecapa-voxceleb", revision="main")
embeddings = extract_speaker_embeddings_from_audios(audios=[audio1, audio2], model=ecapa_model)

```

### Expected Results

```
[tensor([ 16.2103,   7.1482,  15.5266,  17.2746,  16.4431, -20.5747,  13.3308,
         35.0159,  -9.9832, -15.2936,   8.7431, -48.9115, -12.3712,   0.3205,
        -14.1499,  13.8494,  16.3174,  -5.4812, -11.2281,  12.1648, -16.4342,
         -4.4602,  -9.6185,   5.8606,  12.5524,  16.0434,  11.6930, -22.9584,
        -65.9523, -36.1734, -10.4747, -14.4245,  25.6370,  -0.5436, -16.0213,
         -7.1728,  15.8913, -19.5261,  20.3323,   6.4870,   1.3960,  47.1323,
         15.4057,   3.0137,  -6.1211,   9.2284,  -6.8622,  -6.3001,   2.0148,
        -13.6586, -28.2244,  -7.2423,  16.8126,  19.4284,  10.0292,  -0.4392,
        -14.4365, -34.5861,  -7.1497,   9.9081,  -5.9785,  24.1302, -18.4009,
          3.8924,  28.3383, -19.1870, -20.4689, -11.3760, -22.3829, -23.5192,
         -8.0118,  16.7557, -13.0733, -14.4308,  17.9192,  12.4343,  -3.5196,
        -14.0531, -12.2169,  -3.7699,  21.0059, -25.2338,   7.1493,  13.7158,
        -10.7264,   5.1551,  10.4972,  15.6478, -10.8637,  44.2317, -25.2188,
         24.7818,  25.1007, -23.3131, -44.0310,  11.9464,  -4.5547, -17.0638,
         -1.5826,  -1.7229,  29.2818,  -6.5787,   2.9094, -12.4260,  -5.8597,
        -22.0885,  -3.2036,  -6.7363,   7.2173,  -1.7404,  15.0342, -16.2402,
         19.4433,  22.7709,  19.2720, -20.8146,  -8.8956,   9.4334,  -6.2789,
         -1.5953,  -9.3540,  -8.9697, -14.4085, -16.6293, -16.8476,  35.0880,
        -12.9159, -19.0020,  -1.9329,   7.8067,   0.9948,  -4.1761, -16.2834,
          8.2579, -15.5270,  -7.8645,   1.3389,  -9.6198,   3.8484,  -5.8296,
         27.3758,  18.6307,   1.1103,  -7.4372,  12.1074, -13.1453,  12.1791,
          3.2291,  -0.7706, -16.4296,  17.1290, -10.9440,  -1.9359,  42.3513,
         -0.1456, -15.8239, -26.1580,  16.4349, -30.7925,   2.7399,   4.0550,
         18.2988,  -7.7943, -10.2830,  19.3836,  20.8631,  -4.6702,   9.2479,
         13.1044, -13.4122,   4.3439,  -5.9532,  26.1746,  -6.1347,  -9.1914,
...
         13.1044, -13.4122,   4.3439,  -5.9532,  26.1746,  -6.1347,  -9.1914,
        -13.3245,   5.1467,  25.3698,  13.1665,  11.8690, -16.9128,  34.2087,
         30.8796,  -3.3871, -23.9707,  -3.5731,  11.4840,   2.6465, -38.5112,
         -1.7904,  15.8341,  -6.9693])]
```

### Actual Results

```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In[53], [line 5](vscode-notebook-cell:?execution_count=53&line=5)
      [2](vscode-notebook-cell:?execution_count=53&line=2) from senselab.utils.data_structures.model import SpeechBrainModel
      [4](vscode-notebook-cell:?execution_count=53&line=4) ecapa_model = SpeechBrainModel(path_or_uri="speechbrain/spkrec-ecapa-voxceleb", revision="main")
----> [5](vscode-notebook-cell:?execution_count=53&line=5) embeddings = extract_speaker_embeddings_from_audios(audios=[audio1, audio2], model=ecapa_model)

File ~/Documents/git/attempt1/senselab/src/senselab/audio/tasks/speaker_embeddings/api.py:41, in extract_speaker_embeddings_from_audios(audios, model, device)
     [19](https://file+.vscode-resource.vscode-cdn.net/Users/fabiocat/Documents/git/attempt1/senselab/tutorials/~/Documents/git/attempt1/senselab/src/senselab/audio/tasks/speaker_embeddings/api.py:19) """Compute the speaker embedding of audio signals.
     [20](https://file+.vscode-resource.vscode-cdn.net/Users/fabiocat/Documents/git/attempt1/senselab/tutorials/~/Documents/git/attempt1/senselab/src/senselab/audio/tasks/speaker_embeddings/api.py:20) 
     [21](https://file+.vscode-resource.vscode-cdn.net/Users/fabiocat/Documents/git/attempt1/senselab/tutorials/~/Documents/git/attempt1/senselab/src/senselab/audio/tasks/speaker_embeddings/api.py:21) Args:
   (...)
     [38](https://file+.vscode-resource.vscode-cdn.net/Users/fabiocat/Documents/git/attempt1/senselab/tutorials/~/Documents/git/attempt1/senselab/src/senselab/audio/tasks/speaker_embeddings/api.py:38)     torch.Size([192])
     [39](https://file+.vscode-resource.vscode-cdn.net/Users/fabiocat/Documents/git/attempt1/senselab/tutorials/~/Documents/git/attempt1/senselab/src/senselab/audio/tasks/speaker_embeddings/api.py:39) """
     [40](https://file+.vscode-resource.vscode-cdn.net/Users/fabiocat/Documents/git/attempt1/senselab/tutorials/~/Documents/git/attempt1/senselab/src/senselab/audio/tasks/speaker_embeddings/api.py:40) if isinstance(model, SpeechBrainModel):
---> [41](https://file+.vscode-resource.vscode-cdn.net/Users/fabiocat/Documents/git/attempt1/senselab/tutorials/~/Documents/git/attempt1/senselab/src/senselab/audio/tasks/speaker_embeddings/api.py:41)     return SpeechBrainEmbeddings.extract_speechbrain_speaker_embeddings_from_audios(
     [42](https://file+.vscode-resource.vscode-cdn.net/Users/fabiocat/Documents/git/attempt1/senselab/tutorials/~/Documents/git/attempt1/senselab/src/senselab/audio/tasks/speaker_embeddings/api.py:42)         audios=audios, model=model, device=device
     [43](https://file+.vscode-resource.vscode-cdn.net/Users/fabiocat/Documents/git/attempt1/senselab/tutorials/~/Documents/git/attempt1/senselab/src/senselab/audio/tasks/speaker_embeddings/api.py:43)     )
     [44](https://file+.vscode-resource.vscode-cdn.net/Users/fabiocat/Documents/git/attempt1/senselab/tutorials/~/Documents/git/attempt1/senselab/src/senselab/audio/tasks/speaker_embeddings/api.py:44) else:
     [45](https://file+.vscode-resource.vscode-cdn.net/Users/fabiocat/Documents/git/attempt1/senselab/tutorials/~/Documents/git/attempt1/senselab/src/senselab/audio/tasks/speaker_embeddings/api.py:45)     raise NotImplementedError("The specified model is not supported for now.")

File ~/Documents/git/attempt1/senselab/src/senselab/audio/tasks/speaker_embeddings/speechbrain.py:87, in SpeechBrainEmbeddings.extract_speechbrain_speaker_embeddings_from_audios(cls, audios, model, device)
     [81](https://file+.vscode-resource.vscode-cdn.net/Users/fabiocat/Documents/git/attempt1/senselab/tutorials/~/Documents/git/attempt1/senselab/src/senselab/audio/tasks/speaker_embeddings/speechbrain.py:81)         raise ValueError(
     [82](https://file+.vscode-resource.vscode-cdn.net/Users/fabiocat/Documents/git/attempt1/senselab/tutorials/~/Documents/git/attempt1/senselab/src/senselab/audio/tasks/speaker_embeddings/speechbrain.py:82)             "Audio sampling rate " + str(audio.sampling_rate) + 
     [83](https://file+.vscode-resource.vscode-cdn.net/Users/fabiocat/Documents/git/attempt1/senselab/tutorials/~/Documents/git/attempt1/senselab/src/senselab/audio/tasks/speaker_embeddings/speechbrain.py:83)             " does not match expected " + str(expected_sample_rate)
...
---> [87](https://file+.vscode-resource.vscode-cdn.net/Users/fabiocat/Documents/git/attempt1/senselab/tutorials/~/Documents/git/attempt1/senselab/src/senselab/audio/tasks/speaker_embeddings/speechbrain.py:87) waveforms = torch.stack([audio.waveform for audio in audios]).squeeze()
     [89](https://file+.vscode-resource.vscode-cdn.net/Users/fabiocat/Documents/git/attempt1/senselab/tutorials/~/Documents/git/attempt1/senselab/src/senselab/audio/tasks/speaker_embeddings/speechbrain.py:89) # Compute embeddings in a batch
     [90](https://file+.vscode-resource.vscode-cdn.net/Users/fabiocat/Documents/git/attempt1/senselab/tutorials/~/Documents/git/attempt1/senselab/src/senselab/audio/tasks/speaker_embeddings/speechbrain.py:90) embeddings_batch = classifier.encode_batch(waveforms)

RuntimeError: stack expects each tensor to be equal size, but got [1, 78741] at entry 0 and [1, 82581] at entry 1
```

### Additional Notes

@900miles can you please look at this edge case and fix it? thank you!

### PR DESCRIPTION
**Pull request by 900miles**
Adds documentation (and soon a jupyter tutorial) for the speaker embeddings task. Extra attention to the evaluation metrics would be appreciated, this is based on the limited information I could find online.

### CODE REVIEW

File: src/senselab/audio/tasks/speaker_embeddings/api.py [hunk 1]
```diff
@@ -1,4 +1,29 @@
...
+- Detection Cost Function (DCF): A measure of discrimination and calibration in speaker verification systems.
+- Identification Accuracy: The percentage of correctly identified speakers in a closed-set identification task.
+- Diarization Error Rate (DER): Used to evaluate the performance of speaker diarization systems.
+
+### Model Architecture:
+The default model used in this module (speechbrain/spkrec-ecapa-voxceleb) is based on the ECAPA-TDNN architecture,
+which has shown strong performance across various speaker recognition tasks.
+
+**Note**: Performance can vary significantly depending on the specific dataset, task, and evaluation protocol used.
+Always refer to the most recent literature for up-to-date benchmarks.
```
Comment (fabiocat93) [inline src/senselab/audio/tasks/speaker_embeddings/api.py:18]: @900miles, this is great. Thanks. Would you mind adding some references and pointers to other resources that people can check and that you think are relevant, if any?

Comment (fabiocat93) [reply src/senselab/audio/tasks/speaker_embeddings/api.py:18]: e.g., pointers to the papers of the 3 supported models

File: src/senselab/audio/tasks/speaker_embeddings/api.py [hunk 2]
```diff
@@ -1,4 +1,29 @@
...
+
+Speaker embeddings are fixed-dimensional vector representations that capture the unique characteristics of a speaker's
+voice, allowing for tasks such as speaker identification, verification, and diarization.
+
+### Task Overview:
+Speaker embedding extraction is a crucial task in speaker recognition systems. It involves transforming variable-length
+audio signals into fixed-size vector representations that encapsulate speaker-specific information while being robust
+to variations in speech content, background noise, and recording conditions.
+
+### Evaluation Metrics:
```
Comment (fabiocat93) [inline src/senselab/audio/tasks/speaker_embeddings/api.py]: I am not sure these metrics are for speaker embedding extraction. i see them more in the speaker diarization and speaker verification tasks. maybe we can discuss this with the others, but would keep it cleaner and remove this info from here

Comment (900miles) [reply src/senselab/audio/tasks/speaker_embeddings/api.py]: Sounds good. I wasn't sure about metrics for speaker embeddings themselves; from what I could gather they seem to be evaluated by using them for downstream tasks like speaker identification/verification, which is why I included those here. I can remove them for now though

File: src/senselab/audio/tasks/speaker_embeddings/api.py (hunk 1 above)
Comment (900miles) [reply src/senselab/audio/tasks/speaker_embeddings/api.py:18]: We support all speechbrain embeddings models so I've linked to speechbrain and the ecapa-tdnn paper

**Comment by fabiocat93** [comment]
hi @900miles , this is great. I have also fixed the speech_enhancement issue as we said yesterday. Just one more thing before merging this PR: can you please add the visualization functions you use in the tutorial as part of the utility functions for plotting? I think that confusion matrices are used quite frequently and may be helpful to have a well documented function for doing that with all custom params. i haven't seen much the representation you used for speaker embeddings, but I like it and may be helpful at some point. We can also merge this PR and you can open a new one for the new functionalities if you prefer. this is up to you

**Comment by 900miles** [comment]
> i haven't seen much the representation you used for speaker embeddings, but I like it and may be helpful at some point. We can also merge this PR and you can open a new one for the new functionalities if you prefer. this is up to you

Oops! Those visualizations are from claude.ai; I wasn't sure how to best visualize them so this is what Claude suggested. I wasn't too sure about these visualizations so I thought I had deleted them, but it looks like they got pushed somehow! If you like them I can keep them in and document them better, but otherwise my intent was to stop the tutorial after extracting the embeddings, mentioning their use for speaker identification/verification/etc.

I agree confusion matrices would be a good utility for senselab. I will make a separate issue for it.

**Comment by fabiocat93** [comment]
> > i haven't seen much the representation you used for speaker embeddings, but I like it and may be helpful at some point. We can also merge this PR and you can open a new one for the new functionalities if you prefer. this is up to you
> 
> Oops! Those visualizations are from claude.ai; I wasn't sure how to best visualize them so this is what Claude suggested. I wasn't too sure about these visualizations so I thought I had deleted them, but it looks like they got pushed somehow! If you like them I can keep them in and document them better, but otherwise my intent was to stop the tutorial after extracting the embeddings, mentioning their use for speaker identification/verification/etc.

You can keep them and comment it further. I like visuals for effective communication.

> I agree confusion matrices would be a good utility for senselab. I will make a separate issue for it.

Thank you!

**Comment by 900miles** [comment]
@fabiocat93 I pushed a commit that attempts to contextualize the visualization techniques in the tutorial. Is this good? I figured we could merge as-is and then replace the visualization functions with senselab functions when they are implemented.

**Review by fabiocat93 — APPROVED** [review]

**Comment by fabiocat93** [comment]
> @fabiocat93 I pushed a commit that attempts to contextualize the visualization techniques in the tutorial. Is this good? I figured we could merge as-is and then replace the visualization functions with senselab functions when they are implemented.

This is great. Thank you @900miles ! Good job!

**Pull request merged**